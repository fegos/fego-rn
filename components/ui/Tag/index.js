/**
 * Tag组件
 * @author esky 徐达迟 2017/03/30
 */
import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Text,
  StyleSheet,
  TouchableWithoutFeedback,
} from 'react-native';
import { UIComponent } from 'common';
import Icon from '../Icon';

export default class Tag extends UIComponent {
  static defaultProps = {
    defaultChecked: false,
    closeIconName: 'times-circle',
  }
  static propTypes = {
    // 主题类型
    type: PropTypes.string,
    // 尺寸
    size: PropTypes.string,
    // 标签文本，等效于children
    text: PropTypes.node,
    // 标签内容， 此时text无效
    children: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
      PropTypes.node,
    ]),
    // 受控属性：选中状态，需配合onChange使用
    checked: PropTypes.bool,
    // 非控属性：默认选中状态
    defaultChecked: PropTypes.bool,
    // 禁止状态
    disabled: PropTypes.bool,
    // 受控属性：关闭状态，需配合onClose使用
    closed: PropTypes.bool,
    // 关闭按钮图标字体
    closeIconFamily: PropTypes.string,
    // 关闭按钮图标名称
    closeIconName: PropTypes.string,
    onChange: PropTypes.func,
    onClose: PropTypes.func,
  }
  constructor(props) {
    super(props);
    this.state = {
      checked: typeof props.checked === 'boolean' ? props.checked : props.defaultChecked,
    };
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.checked !== nextProps.checked) {
      return ({
        checked: nextProps.checked,
      });
    }
    return null;
  }
  _onCloseClick = () => {
    const { onClose } = this.props;
    if (typeof onClose === 'function') {
      onClose(!this.state.closed);
    }
  }
  _onClick = () => {
    const { onChange, checked } = this.props;
    const _checked = !this.state.checked;
    if (typeof checked !== 'boolean') {
      this.setState({
        checked: _checked,
      });
    }
    if (typeof onChange === 'function') {
      onChange(_checked, this);
    }
  }
  _renderCloseBtn = () => {
    const { style } = this;
    const { closed, closeIconFamily, closeIconName } = this.props;
    // 未启用关闭功能
    if (typeof closed !== 'boolean') return null;
    return (
      <TouchableWithoutFeedback onPress={this._onCloseClick}>
        <View style={style.close}>
          <Icon family={closeIconFamily} name={closeIconName} style={style.closeIcon} />
        </View>
      </TouchableWithoutFeedback>
    );
  }
  _renderInner() {
    const { style } = this;
    const { text, children, disabled } = this.props;
    let inner = children || text;
    if (typeof inner === 'string') {
      inner = (
        <Text style={[
          style.text,
          this.state.checked ? style.textChecked : {},
          disabled ? style.textdisabled : {}]}
        >
          {inner}
        </Text>
      );
    }
    return inner;
  }

  render() {
    const { style } = this;
    const { disabled, closed } = this.props;
    const { checked } = this.state;
    const innerEl = this._renderInner();
    const colseEl = this._renderCloseBtn();
    // 已关闭
    if (typeof closed === 'boolean' && closed) {
      return null;
    }
    // 禁止状态
    if (disabled) {
      return (
        <View style={[style.container, style.disabled]}>
          {innerEl}
          {colseEl}
        </View>
      );
    }
    // 正常状态
    return (
      <TouchableWithoutFeedback onPress={this._onClick}>
        <View style={[style.container, checked ? style.checked : {}]}>
          {innerEl}
          {colseEl}
        </View>
      </TouchableWithoutFeedback>
    );
  }
}

Tag.baseStyle = {
  container: {
    flexDirection: 'row',
    overflow: 'visible',
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 3,
    paddingVertical: 5,
    paddingHorizontal: 5,
    alignItems: 'center',
    borderColor: '#bbb',
    backgroundColor: '#FFF',
  },
  disabled: {
    borderColor: '#ddd',
    backgroundColor: '#ddd',
  },
  checked: {
    borderColor: '#108ee9',
    backgroundColor: '#FFF',
  },
  text: {
    fontSize: 12,
    color: '#888',
  },
  textChecked: {
    color: '#108ee9',
  },
  textDisabled: {
    color: '#ddd',
  },
  close: {
    paddingLeft: 5,
  },
  closeIcon: {
    color: '#bbb',
    fontSize: 14,
  },
};

Tag.themeStyle = {
  size: {
    small: {
      container: {
        paddingVertical: 2,
        paddingHorizontal: 2,
      },
      text: {
        fontSize: 10,
      },
      close: {
        paddingLeft: 2,
      },
      closeIcon: {
        fontSize: 11,
      },
    },
    large: {
      container: {
        paddingVertical: 7,
        paddingHorizontal: 7,
      },
      text: {
        fontSize: 16,
      },
      close: {
        paddingLeft: 7,
      },
      closeIcon: {
        fontSize: 18,
      },
    },
  },
};
