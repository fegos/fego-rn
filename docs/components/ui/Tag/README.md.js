webpackJsonp([6],{496:function(t,a){t.exports={content:["article",["h1","\u6807\u7b7e"],["h3","\u7ec4\u4ef6\u63cf\u8ff0"],["ul",["li",["p","\u53ef\u9009\u62e9\uff0c\u53ef\u5173\u95ed\u7684\u6807\u7b7e\u5143\u7d20"],["h3","\u793a\u4f8b\u4ee3\u7801"]]],["pre",{lang:"html",highlighted:'<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Tag</span><span class="token punctuation">></span></span>\u666e\u901aTag<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Tag</span><span class="token punctuation">></span></span>\n<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Tag</span> <span class="token attr-name">defaultSelected</span><span class="token punctuation">></span></span>\u9ed8\u8ba4\u5df2\u9009Tag<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Tag</span><span class="token punctuation">></span></span>\n<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Tag</span> <span class="token attr-name">disabled</span><span class="token punctuation">></span></span>\u7981\u7528Tag<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Tag</span><span class="token punctuation">></span></span>\n&lt;Tag onClose={()=>doSomething()}>\u7981\u7528Tag<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Tag</span><span class="token punctuation">></span></span>\n&lt;Tag onChange={()=>doSomething()}>\u7981\u7528Tag<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Tag</span><span class="token punctuation">></span></span>'},["code","<Tag>\u666e\u901aTag</Tag>\n<Tag defaultSelected>\u9ed8\u8ba4\u5df2\u9009Tag</Tag>\n<Tag disabled>\u7981\u7528Tag</Tag>\n<Tag onClose={()=>doSomething()}>\u7981\u7528Tag</Tag>\n<Tag onChange={()=>doSomething()}>\u7981\u7528Tag</Tag>"]],["h2","API"],["h3","props\u5217\u8868"],["table",["thead",["tr",["th","\u5c5e\u6027"],["th","\u8bf4\u660e"],["th","\u7c7b\u578b"],["th","\u9ed8\u8ba4\u503c"]]],["tbody",["tr",["td","type"],["td","\u4e3b\u9898\u7c7b\u578b"],["td","string"],["td","default"]],["tr",["td","size"],["td","\u5c3a\u5bf8"],["td","string"],["td","\u53ef\u9009\u62e9small,large"]],["tr",["td","text"],["td","\u6807\u7b7e\u663e\u793a\u7684\u6587\u672c"],["td","-"],["td","-"]],["tr",["td","children"],["td","\u6807\u7b7e\u663e\u793a\u7684\u6587\u672c\uff0c\u4f7f\u7528\u65f6 text \u5c5e\u6027\u5931\u6548"],["td","-"],["td","-"]],["tr",["td","selected"],["td","\u662f\u5426\u9009\u62e9"],["td","boolean"],["td","false"]],["tr",["td","defaultSelected"],["td","\u662f\u5426\u9ed8\u8ba4\u9009\u62e9"],["td","boolean"],["td","false"]],["tr",["td","disabled"],["td","\u662f\u5426\u7981\u7528"],["td","boolean"],["td","false"]],["tr",["td","closed"],["td","\u662f\u5426\u53ef\u4ee5\u5173\u95ed"],["td","boolean"],["td","false"]],["tr",["td","onClose"],["td","\u5173\u95ed\u540e\u7684\u56de\u8c03"],["td","func"],["td"]],["tr",["td","onChange"],["td","\u70b9\u51fb\u540e\u7684\u56de\u8c03"],["td","func"],["td"]],["tr",["td","closeIconName"],["td","\u5173\u95ed\u6309\u94ae\u56fe\u6807\u540d\u79f0"],["td","string"],["td"]],["tr",["td","closeIconFamily"],["td","\u5173\u95ed\u6309\u94ae\u56fe\u6807\u5b57\u4f53\u5e93"],["td","string"],["td"]]]],["h3","\u6837\u5f0f\u5bf9\u8c61styles"],["table",["thead",["tr",["th","\u5c5e\u6027"],["th","\u8bf4\u660e"],["th","\u9002\u7528\u7c7b\u578b"]]],["tbody",["tr",["td","container"],["td","Tag\u5bb9\u5668\u6837\u5f0f"],["td","View"]],["tr",["td","text"],["td","Tag\u6587\u672c\u6837\u5f0f"],["td","Text"]],["tr",["td","textSelected"],["td","\u5df2\u9009Tag\u6587\u672c\u6837\u5f0f"],["td","Text"]],["tr",["td","textDisabled"],["td","\u7981\u7528Tag\u6587\u672c\u6837\u5f0f"],["td","Text"]]]]],meta:{title:"Tag",subTitle:"\u6807\u7b7e",filename:"components/ui/Tag/README.md"},toc:["ul",["li",["a",{className:"bisheng-toc-h1",href:"#\u6807\u7b7e",title:"\u6807\u7b7e"},"\u6807\u7b7e"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u7ec4\u4ef6\u63cf\u8ff0",title:"\u7ec4\u4ef6\u63cf\u8ff0"},"\u7ec4\u4ef6\u63cf\u8ff0"]],["li",["a",{className:"bisheng-toc-h2",href:"#API",title:"API"},"API"]],["li",["a",{className:"bisheng-toc-h3",href:"#props\u5217\u8868",title:"props\u5217\u8868"},"props\u5217\u8868"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u6837\u5f0f\u5bf9\u8c61styles",title:"\u6837\u5f0f\u5bf9\u8c61styles"},"\u6837\u5f0f\u5bf9\u8c61styles"]]]}}});