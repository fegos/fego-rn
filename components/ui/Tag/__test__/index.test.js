import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Tag from '../index';

describe('Tag Tests', () => {
  it('renders correctly', () => {
    const tree = Renderer.create((
      <Tag text="这是一个tag" />
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('defaultSelected state', () => {
    const wrapper = shallow(<Tag>I’m a tag</Tag>);
    expect(wrapper.instance().state.checked).toBe(false);

    wrapper.simulate('press');
    expect(wrapper.instance().state.checked).toBe(true);
  });

  it('press event', () => {
    const handler = jest.fn();
    const wrapper = shallow(<Tag onPress={handler}>I’m a tag too</Tag>);

    expect(wrapper.instance().state.checked).toBe(false);
    wrapper.simulate('press');
    expect(wrapper.instance().state.checked).toBe(true);
  });
});
