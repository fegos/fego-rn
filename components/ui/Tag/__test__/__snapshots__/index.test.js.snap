// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Tests renders correctly 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderColor": "#bbb",
        "borderRadius": 3,
        "borderWidth": 0.5,
        "flexDirection": "row",
        "overflow": "visible",
        "paddingHorizontal": 5,
        "paddingVertical": 5,
      },
      Object {},
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#888",
          "fontSize": 12,
        },
        Object {},
        Object {},
      ]
    }
  >
    这是一个tag
  </Text>
</View>
`;
